### DISCUSSION TITLE
[ld65] Generating relocations --- how?

### DISCUSSION THREAD

**Discussion by davidgiven** [cc65/cc65#1860] [General]
I'm trying to add support for a binary format which has a relocation table. The general workflow is that the binary is linked at a particular address, and then fixed up at load time by adjusting any addresses inside the program. So, I need to collect a list of addresses which need fixing up.

Inside a binary emitter, I iterate the list of memory areas (one, for me); then I iterate the sections within the area; then, I call `SegWrite` to emit the code. This calls me back to emit actual expressions. (This was all cut-and-pasted from the other emitters.)

However, I'm at a loss of what to do next. I was assuming that the expressions would have been evaluated by now, so I'd just need to check the size of the value being written and whether it's pointing at a movable memory area; but because expressions are complex it's not that easy. I've tried calling `GetSegExprVal` to extract the Section of the resulting value but I keep getting TooComplex set.

The only thing I can think of is to just call `GetValue` multiple times, adjusting the memory area base address between times, and seeing if the value changes. But I don't want to do that because it's horrible. Is there an easier way?

**Comment by Marrin** [comment]
Have you looked at the o65 output format?

**Reply by davidgiven** [reply]
Sure, but O65 doesn't work for my target platform.

**Comment by Marrin** [comment]
But it is a format with a relocation table, so you can look how that was implemented.  I thought that may help.

**Reply by davidgiven** [reply]
Oh, I see --- sorry, misunderstood you. O65's relocation tables work completely differently, unfortunately. It seems to serialise the expression to then be evaluated by whatever reads it. However I'm kinda puzzled as to _why_ it works as it's got its own copy of `GetSegExprVal` which also appears to fail with TooComplex errors, but then it goes ahead and uses the values anyway, so I'm not sure I've correctly understood the code.
